File: kits/python/energium/position.py
from .game_constants import DIRECTIONS, GAME_CONSTANTS
import math
class Position:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y

    def equals(self, opos):
        return self.x == opos.x and self.y == opos.y

    def is_adjacent(self, pos):
        """
        returns true if this position is adjacent to pos
        """
        dx = self.x - pos.x
        dy = self.y - pos.y
        if (abs(dx) + abs(dy) > 1):
            return False
        return True
    def translate(self, direction, units):
        """
        translate a position in a direction by some units and returns a translated
        Position
        """
        if direction == DIRECTIONS.NORTH:
            return Position(self.x, self.y - units)
        elif direction == DIRECTIONS.EAST:
            return Position(self.x + units, self.y)
        elif direction == DIRECTIONS.SOUTH:
            return Position(self.x, self.y + units)
        elif direction == DIRECTIONS.WEST:
            return Position(self.x - units, self.y)

    def distance_to(self, pos):
        """
        returns euclidean distance to the pos from this position
        """
        dx = pos.x - self.x
        dy = pos.y - self.y
        return math.sqrt(dx * dx + dy * dy)

    def direction_to(self, targetPos):
        """
        gives direction that moves closest to targetPos from this position or None if staying put is closer
        """
        checkDirections = [
            DIRECTIONS.NORTH,
            DIRECTIONS.EAST,
            DIRECTIONS.SOUTH,
            DIRECTIONS.WEST,
        ]
        closestDirection = None
        closestDist = self.distance_to(targetPos)
        for dir in checkDirections:
            newpos = self.translate(dir, 1)
            dist = targetPos.distance_to(newpos)
            if (dist < closestDist):
                closestDist = dist
                closestDirection = dir
        return closestDirection

    def taxicab_distance_to(self, pos):
        """
        returns taxicab distance to the pos from this position
        """
        return abs(self.x - pos.x) + abs(self.y - pos.y)

    def direction_to_quick(self, target):
        """
        Same as direction_to but runs quicker.
        """
        closest_direction = None
        taxicab_dist = taxicab_distance(self, target)
        bigger_side_length = max(abs(self.x - target.x), abs(self.y - target.y))
        for dir in [DIRECTIONS.NORTH, DIRECTIONS.EAST,
                    DIRECTIONS.SOUTH, DIRECTIONS.WEST]:
            newpos = self.translate(dir, 1)
            dist = taxicab_distance(newpos, target)
            if dist < taxicab_dist:
                taxicab_dist = dist
                bigger_side_length = max(abs(newpos.x - target.x), abs(newpos.y - target.y))
                closest_direction = dir
            elif dist == taxicab_dist:
                s = max(abs(newpos.x - target.x), abs(newpos.y - target.y))
                if s < bigger_side_length:
                    bigger_side_length = s
                    closest_direction = dir
        return closest_direction

File: kits/java/energium/Position.java
package energium;

public class Position {
  public int x;
  public int y;

  public Position(int x, int y) {
    this.x = x;
    this.y = y;
  }

  /** Returns true if this position is adjacent to another position */
  public boolean isAdjacent(Position pos) {
    int dx = this.x - pos.x;
    int dy = this.y - pos.y;
    if (Math.abs(dx) + Math.abs(dy) > 1) {
      return false;
    }
    return true;
  }

  public boolean equals(Position pos) {
    return this.x == pos.x && this.y == pos.y;
  }

  /**
   * Translate a position in a direction by some units and returns a translated
   * Position
   */
  public Position translate(Direction direction, int units) {
    switch (direction) {
      case NORTH:
        return new Position(this.x, this.y - units);
      case EAST:
        return new Position(this.x + units, this.y);
      case SOUTH:
        return new Position(this.x, this.y + units);
      case WEST:
        return new Position(this.x - units, this.y);
    }
    throw new Error("Did not supply valid direction");
  }

  /** Returns distance to pos from this position */
  public double distanceTo(Position pos) {
    int dx = pos.x - this.x;
    int dy = pos.y - this.y;
    return Math.sqrt(dx * dx + dy * dy);
  }

  /** Returns closest direction to targetPos, or null if staying put is best */
  public Direction directionTo(Position targetPos) {
    Direction[] checkDirections = { Direction.NORTH, Direction.EAST, Direction.SOUTH, Direction.WEST, };
    Direction closestDirection = null;
    double closestDist = this.distanceTo(targetPos);
    for (Direction dir : checkDirections) {
      Position newpos = this.translate(dir, 1);
      double dist = targetPos.distanceTo(newpos);
      if (dist < closestDist) {
        closestDist = dist;
        closestDirection = dir;
      }
    }
    return closestDirection;
  }

  @Override
  public String toString() {
    return "(" + this.x + ", " + this.y + ")";
  }
}
